internal/delete/delete_item_action.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2020 the Velero contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package delete
+
+import (
+	"io"
+
+	"github.com/sirupsen/logrus"
+	"github.com/vmware-tanzu/velero/pkg/discovery"
+	"github.com/vmware-tanzu/velero/pkg/plugin/velero"
+	"github.com/vmware-tanzu/velero/pkg/util/collections"
+	"github.com/vmware-tanzu/velero/pkg/util/filesystem"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+// Context provides the necessary environment to run DeleteItemAction plugins
+type Context struct {
+	BackupReader    io.Reader
+	Actions         []velero.DeleteItemAction
+	Filesystem      filesystem.Interface
+	Log             logrus.FieldLogger
+	DiscoveryHelper discovery.Helper
+}
+
+func InvokeDeleteActions(ctx *Context) error {
+	// helper will need to be passed in to the controller.
+	resolvedActions, err := resolveActions(ctx.Actions, ctx.DiscoveryHelper)
+
+	// No actions installed and no error means we don't have to continue;
+	// just do the backup deletion without worrying about plugins.
+	if resolvedActions == nil && err == nil {
+		ctx.Log.Debug("No delete item actions present, proceeding with rest of backup deletion process")
+		return nil
+	}
+
+	// // get items out of backup tarball into a temp directory
+	// dir, err := archive.NewExtractor(ctx.Log, ctx.Filesystem).UnzipAndExtractBackup(ctx.BackupReader)
+	// if err != nil {
+	// 	return errors.Wrapf(err, "error extracting backup")
+
+	// }
+	// defer ctx.Filesystem.RemoveAll(dir)
+
+	// backupResources, err := archive.NewParser(ctx.Log, ctx.Filesystem).Parse(dir)
+	// if err != nil {
+	// 	return errors.Wrapf(err, "error parsing backup contents")
+	// }
+	// // iterate through resources. Need to transform these from a map into a string, probably with some ordering like in restore.
+	// for _, resource := range backupResources {
+	// 	gvr, _, err := ctx.DiscoveryHelper.ResourceFor(schema.ParseGroupKind(resource).WithVersion(""))
+	// 	if err != nil {
+	// 		return errors.Wrapf(err, "failed to resolve resource into complete group/version/resource: %v", resource)
+	// 	}
+
+	// }
+	// unmarshal items into Unstructured
+	// get applicable actions for the group
+	// apply actions to the Unstructured items
+
+	return nil
+}
+
+type resolvedAction struct {
+	velero.DeleteItemAction
+
+	resourceIncludesExcludes  *collections.IncludesExcludes
+	namespaceIncludesExcludes *collections.IncludesExcludes
+	selector                  labels.Selector
```

Comment:
suggest creating a new struct, with just these three fields, that can be used in all ItemAction plugins.
That should save us a lot of code duplication across the plugin types, and also perhaps the hooks.

Reply:
This will avoid writing `resolveActions` for every plugin.

Reply:
we can introduce the new type now and use it in the DeleteItemAction plugin and update the other plugin types as a follow-up. WDYT?

Reply:
Yeah, I want to unify all of them.

The problem I ran into is that each plugin has a slightly different signature, and this point was missed in the design phase.

There's a lot of duplication between restoration and deletion that I'd like to reduce, though deletion is much simpler as it turns out.